src/frontCANnode/src/input.c

```diff
@@ -0,0 +1,178 @@
+  ADC_Input_T *adc = input->adc;
+  uint32_t next_updated = adc->last_updated + ADC_UPDATE_PERIOD_MS;
+
+  if (next_updated < input->msTicks) {
+    adc->accel_1_raw = ADC_Read(ACCEL_1_CHANNEL);
+    adc->accel_2_raw = ADC_Read(ACCEL_2_CHANNEL);
+    adc->brake_1 = ADC_Read(BRAKE_1_CHANNEL);
+    adc->brake_2 = ADC_Read(BRAKE_2_CHANNEL);
+    adc->steering_pot = ADC_Read(STEERING_CHANNEL);
+    adc->last_updated = input->msTicks;
+
+    adc->accel_1 = transform1(adc->accel_1_raw);
+    adc->accel_2 = transform2(adc->accel_2_raw);
+  }
+}
+
+uint32_t click_time_to_mRPM(uint32_t us_per_click) {
+  const float us_per_rev = us_per_click * 1.0 * NUM_TEETH;
+
+  const float s_per_rev = us_per_rev / MICROSECONDS_PER_SECOND_F;
+
+  const float mrev_per_s = MILLIREVS_PER_REV_F / s_per_rev;
+
+  const float mrev_per_min = mrev_per_s * SECONDS_PER_MINUTE;
+  return (uint32_t)mrev_per_min;
+}
+
+void update_wheel_speed(Speed_Input_T *speed, uint32_t msTicks) {
+  if (speed->last_speed_read_ms + WHEEL_SPEED_READ_PERIOD_MS < msTicks) {
+    // Capture values
+    speed->last_speed_read_ms = msTicks;
+    uint8_t wheel;
+    for(wheel = 0; wheel < NUM_WHEELS; wheel++) {
+      const uint32_t count = speed->num_ticks[wheel];
+      uint8_t idx;
+      if (count > 0) {
+        // If there are x ticks so far, the last tick is index (x - 1)
+        idx = (count - 1) % NUM_TEETH;
+      } else {
+        idx = 0;
+      }
+
+      uint32_t moving_avg;
+      if (count < NUM_TEETH) {
+        moving_avg = 0;
+      } else {
+        const uint32_t avg = speed->big_sum[wheel] / SUM_ALL_TEETH;
+        moving_avg = avg;
+      }
+      const bool timeout =
+        speed->last_updated[wheel] + WHEEL_SPEED_TIMEOUT_MS < msTicks;
+      speed->wheel_stopped[wheel] = timeout || count == 0;
+      speed->disregard[wheel] = timeout;
+
+      // Save value
+      uint32_t *ptr;
+      if (wheel == LEFT) {
+        ptr = &speed->front_left_wheel_speed;
+      } else if (wheel == RIGHT) {
+        ptr = &speed->front_right_wheel_speed;
+      } else continue;
+      if (speed->wheel_stopped[wheel]) {
+        *ptr = 0;
+        continue;
+      }
+      if (count < NUM_TEETH) {
+        *ptr = click_time_to_mRPM(speed->last_tick[wheel][idx]);
+      } else {
+        *ptr = click_time_to_mRPM(moving_avg);
+      }
+    }
+  }
+}
+
+uint16_t linear_transfer_fn(uint32_t reading, uint16_t desired_width, uint16_t lower_bound, uint16_t upper_bound) {
+  // Mostly copied from MY17
+
+  // Ensure reading is within expected range
+  if (lower_bound > reading) reading = lower_bound;
+  if (upper_bound < reading) reading = upper_bound;
+
+  // Make reading between 0 and diff
+  const uint16_t diff = upper_bound - lower_bound;
+  reading = reading - lower_bound;
+
+  // Now scale from [0:diff] to [0:desired_width].
+  // Note: it's critical that reading be a 32 bit int because otherwise this line will cause overflow!
+  reading = reading * desired_width;
+  reading = reading / diff;
```

Comment:
Should it be done with no potential overflow at the price of a little less accuracy?